src/run-tracker.ts

```diff
@@ -0,0 +1,143 @@
+import type { ActorRun } from 'apify-client';
+
+import type { RunInfo, UpdateCallback } from './types.js';
+import { isRunFailStatus } from './utils/apify-client.js';
+import { getRunUrl } from './utils/apify-console.js';
+import type { GlobalContext } from './utils/context.js';
+import type { Storage } from './utils/storage.js';
+
+const TRACKED_RUNS_KEY = 'RUNS';
+
+type RunInfoRecord = { [runName: string]: RunInfo };
+
+interface TrackedRuns {
+    current: RunInfoRecord;
+    failedHistory: { [runName: string]: RunInfo[] };
+}
+
+export interface RunTrackerOptions {
+    storage?: Storage;
+    storagePrefix?: string;
+    onUpdate?: UpdateCallback;
+}
+
+export class RunTracker {
+    private constructor(
+        private readonly context: GlobalContext,
+        private readonly trackedRuns: TrackedRuns,
+        private readonly onUpdate?: UpdateCallback,
+    ) {
+        this.itemsChangedCallback();
+    }
+
+    static async new(context: GlobalContext, options?: RunTrackerOptions): Promise<RunTracker> {
+        const defaultTrackedRuns = getDefaultTrackedRuns();
+        const storageKey = `${options?.storagePrefix ?? ''}${TRACKED_RUNS_KEY}`;
+        const trackedRuns =
+            (await options?.storage?.useState<TrackedRuns>(storageKey, defaultTrackedRuns)) ?? defaultTrackedRuns;
```

Comment:
And we are guaranteed that it is the same underlying object which is returned, so even if we create two of these they will always be synchronized...